### PR TITLE
feat: Air gap EC2 and Sagemaker workspaces

### PR DESCRIPTION
**Pull request by nguyen102**
Issue #, if available:

Description of changes:

Removed external dependencies and put them into the `environment-files` folder in S3.

### These are the dependencies placed into the S3 folder.

Dependencies for Sagemaker and Ec2-Linux
* jq-1.5
* goofys

Dependencies for Sagemaker
* fuse-2.9.4
* autostop.py
* 
Dependencies for EC2-Linux
* ec2-instance-connect
* fuse-2.9.2

#### EC2-Linux and EC2-Windows requires access to Cloudformation Private Endpoint to signal that they have successfully finished their `goofys` setup script. This allows Cloudformation to mark their respective stack as `Successfully Created`

To allow these respective workspace access to Cloudformation Endpoint, we need these workspaces to set their Security group outbound rules to point to the CF Endpoint's security group. We also need the CF Endpoint to set their Security group inbound rules to be the workspace SG.

To accomplish this we've set up this network topology.

EC2 Linux/ EC2 Windows Workspace has `DefaultSG` and`CustomSG` attached to it
* Outbound Rule of CustomSG: `CFN SG`

CFN Endpoint has `CFN SG` attached to it
* Inbound Rule of CFN SG: `DefaultSG`

By attaching the `DefaultSG` to the workspaces we can have CFN endpoint point to the workspaces when the CFN Endpoint is initially created.

#### To allow AppStream to connect to ec2 instances
EC2 Linux/ EC2 Windows Workspace has `DefaultSG` and`CustomSG` attached to it
* Inbound Rule of CustomSG: `AppStreamSG`

AppStream has `AppStreamSG` attached to it
* Outbound Rule: `DefaultSG`

By attaching the `DefaultSG` to the workspaces we can have AppStream allow access to the workspaces when AppStream resources are initially created during `onboard-account`.

We export `CFN Security Group`, `Default SG`, and `AppStreamSG` in the Cloudformation stack to make it easy for the workspaces to consume these variables.

 ### Testing
 Manually created new Cloudformation stacks with ec2-window, ec2-linux, and sagemaker Cloudformation templates. These stacks were placed in the Private AppStream Workspace. I then logged into each of these workspaces and tested that they didn't have internet access but were able to access S3 studies.

Checklist:

- [x] Have you successfully deployed to an AWS account with your changes?

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-linux-instance.cfn.yml [hunk 1]
```diff
@@ -133,6 +136,9 @@ Resources:
           FromPort: -1
           ToPort: -1
           CidrIp: !Ref AccessFromCIDRBlock
+        - DestinationSecurityGroupId:
+            Fn::ImportValue: !Sub "${SolutionNamespace}-CfnEndpointSecurityGroup"
+          IpProtocol: '-1'
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-linux-instance.cfn.yml]: Don't we need isAppStream flag? Does this work without that flag?

File: scripts/check-dependency-vulnerabilities.sh [hunk 2]
```diff
@@ -7,7 +7,7 @@ pnpm audit --json --audit-level high --prod > audit-results.json
 high_vul=$(cat audit-results.json | jq '.metadata.vulnerabilities.high')
 critical_vul=$(cat audit-results.json | jq '.metadata.vulnerabilities.critical')
 
-high_vul_threshold=2
+high_vul_threshold=4
```
Comment (jn1119) [inline scripts/check-dependency-vulnerabilities.sh]: why was this changed?

Comment (nguyen102) [reply scripts/check-dependency-vulnerabilities.sh]: Ah yes, so it looks like `prismjs` requires an update to version `2.4.0` but we're using older version of `prismjs` in `base-raas-ui`. That is why that check is failing. This affects our repo as a whole and is not directly related to this PR. 

This [PR](https://github.com/awslabs/service-workbench-on-aws/pull/556) fixes it for our `docs` subdirectory, but not our `base-raas-ui` directory.

File: addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-windows-instance.cfn.yml [hunk 3]
```diff
@@ -155,6 +158,9 @@ Resources:
           FromPort: -1
           ToPort: -1
           CidrIp: !Ref AccessFromCIDRBlock
+        - DestinationSecurityGroupId:
```
Comment (nguyen102) [inline addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-windows-instance.cfn.yml]: Note: Add conditional on AppStream flag

File: addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/provision-environment/provision-environment.js [hunk 4]
```diff
@@ -141,6 +141,15 @@ class ProvisionEnvironment extends StepBase {
...
       addParam('KeyName', keyName);
     }
+
+    if (type === 'ec2-windows' || type === 'ec2-linux') {
+      addParam('SolutionNamespace', this.settings.get('solutionNamespace'));
+    }
+
+    if (type === 'sagemaker') {
+      addParam('IsAppStreamEnabled', this.settings.get('isAppStreamEnabled').toString());
+    }
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/provision-environment/provision-environment.js]: is there a possibility that these parameters will be added to an old template and not to the new template that customers haven't imported yet. Can we fail environment provisioning in case the template is old and not suited for AppStream but the solution has it enabled?

File: addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-linux-instance.cfn.yml (hunk 1 above)
Comment (nguyen102) [reply addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-linux-instance.cfn.yml]: Good catch. Adding the flag.

File: addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/provision-environment/provision-environment.js (hunk 4 above)
Comment (nguyen102) [reply addons/addon-base-raas/packages/base-raas-workflow-steps/lib/steps/provision-environment/provision-environment.js]: Turns out this `provision-environment` file was used for built in workspaces and not SC workspaces. I made the updates in the correct file, which is `env-config-var-service`. I tested this, and even if we pass these parameters to an old template, the workspace still gets started correctly. 

In terms of checking whether an environment is launching the right type of workspace, I think that check should be done earlier in the process than when the workflowLoopRunner is started. When a user switch from normal workspaces to AppStream enabled workspace, we should have a post deployment script that disables all of the "normal" workspace types. And then from then on, we should check that a workspace type is AppStream compatible before letting the user approves the workspace for launch.

File: addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/onboard-account.cfn.yml [hunk 5]
```diff
@@ -734,4 +757,16 @@ Outputs:
     Condition: isAppStream
     Value: !Ref AppStreamStack
 
+  CFNEndpointSG:
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/onboard-account.cfn.yml:767]: Where is the ingress rule for this defined? Do we not need it for EC2 instances to connect to CFN?

File: addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-linux-instance.cfn.yml [hunk 6]
```diff
@@ -173,6 +187,10 @@ Resources:
           DeviceIndex: '0'
           GroupSet:
             - !Ref SecurityGroup
+            - !If
+              - AppStreamEnabled
+              - Fn::ImportValue: !Sub "${SolutionNamespace}-SwbVPCDefaultSG"
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-linux-instance.cfn.yml:198]: Does this give each EC2 instance (Windows or Linux based) access to other Windows/Linux systems as well? The instances will still have ssh key/password protection etc but I guess those instances will be reachable? Asking this because default's security groups default ingress might allow itself.

Comment (nguyen102) [reply addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-linux-instance.cfn.yml:198]: Yep, that's right. Since default security group has ingress to itself other EC2 instances have connectivity to each other. However, since those instances still require ssh key, those instances are still secured.

File: addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/onboard-account.cfn.yml (hunk 5 above)
Comment (nguyen102) [reply addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/onboard-account.cfn.yml:767]: Oops, updating the PR to include that now. It'll be on line 601 to 603.

File: addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-windows-instance.cfn.yml [hunk 7]
```diff
@@ -155,11 +163,23 @@ Resources:
...
         - IpProtocol: tcp
           FromPort: 3389
           ToPort: 3389
           CidrIp: !Ref AccessFromCIDRBlock
+        - !If
+          - AppStreamEnabled
+          - SourceSecurityGroupId:
+              Fn::ImportValue: !Sub "${SolutionNamespace}-SwbAppStreamSG"
+            IpProtocol: '-1'
+          - !Ref "AWS::NoValue"
```
Comment (jn1119) [inline addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-windows-instance.cfn.yml:182]: Can you add similar egress and ingress rules to RStudio as well. Similarly default security group needs to be added in NetworkInterfaces there.

Comment (nguyen102) [reply addons/addon-base-raas/packages/base-raas-cfn-templates/src/templates/service-catalog/ec2-windows-instance.cfn.yml:182]: updated

**Review by jn1119 — APPROVED** [review]
